File: ganilib-core/src/main/java/com/glib/core/ui/layout/GLayoutParams.java
package com.glib.core.ui.layout;

import android.widget.LinearLayout;

import com.glib.core.ui.style.Length;

public class GLayoutParams<T extends GLayoutParams> extends LinearLayout.LayoutParams {
  public GLayoutParams() {
    super(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
  }

  private T self() {
    return (T) this;
  }

  public T width(int width) {
    this.width = Length.INSTANCE.dpToPx(width);
    return self();
  }

  public T height(int height) {
    this.height = Length.INSTANCE.dpToPx(height);
    return self();
  }

  public T margins(int left, int top, int right, int bottom){
    super.setMargins(Length.INSTANCE.dpToPx(left), Length.INSTANCE.dpToPx(top), Length.INSTANCE.dpToPx(right), Length.INSTANCE.dpToPx(bottom));
    return self();
  }
}

File: glib-core/src/main/java/com/glib/core/ui/layout/GLayoutParams.kt
package com.glib.core.ui.layout

import android.widget.LinearLayout
import com.glib.core.ui.style.Length.dpToPx

class GLayoutParams<T : GLayoutParams<T>> : LinearLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT) {
    private fun self(): T {
        return this as T
    }

    fun width(width: Int): T {
        this.width = dpToPx(width)
        return self()
    }

    fun height(height: Int): T {
        this.height = dpToPx(height)
        return self()
    }

    fun margins(left: Int, top: Int, right: Int, bottom: Int): T {
        super.setMargins(dpToPx(left), dpToPx(top), dpToPx(right), dpToPx(bottom))
        return self()
    }
}
